charm/kafka/reactive/kafka.py

```diff
@@ -199,24 +180,23 @@ def storage_attach():
     log_dir = os.path.join(mount, "logs")
     unitdata.kv().set('kafka.storage.log_dir', log_dir)
     hookenv.log('Kafka logs storage attached at {}'.format(log_dir))
+
     # Stop Kafka; removing the kafka.started state will trigger
     # a reconfigure if/when it's ready
-    kafka = Kafka()
-    kafka.close_ports()
-    kafka.stop()
-    remove_state('kafka.started')
-    hookenv.status_set('waiting', 'reconfiguring to use attached storage')
-    set_state('kafka.storage.logs.attached')
+    remove_state('kafka.configured')
+    set_flag('kafka.force-reconfigure')
 
 
 @hook('logs-storage-detaching')
 def storage_detaching():
     unitdata.kv().unset('kafka.storage.log_dir')
-    kafka = Kafka()
-    kafka.close_ports()
-    kafka.stop()
```

Comment:
Kafka needs to be stopped here.

Reply:
Definitely needs to be stopped here immediately, because we need to terminate all the processes that are accessing files on the volume, so it can be unmounted. Otherwise that umount can hang.